Bundle SSL certs with the templates

Faless opened the pull request:
If this is undesired it can be avoided by specifying `builtin_certs=no`.
Bundled SSL certs will be used unless you specify an override in:
`Project Settings -> SSL -> Certificates`.

Closes #17328 .

Faless committed:
Bundle SSL certs with the templates.

If this is undesired it can be avoided by specifying builtin_certs=no .
Bundled SSL certs will be used unless you specify an override in:
Project Settings -> SSL -> Certificates .

akien-mga commented:
That's a good idea, but at the same time I think we need to provide a way to use system certs if `builtin_certs=no`. I've actually discussed this recently with openSUSE packagers who are trying to unbundle certs in Godot to use the system provided ones, for security reasons (if system certs are updated/configured in a specific way, Godot as a packaged application is supposed to respect that - currently you'd have to do a Godot update to remove now-untrusted certs for example).

See the patches in https://build.opensuse.org/package/show/home:cunix:branches:games:tools/godot
Note that those patches are for 3.0.x, so OpenSSL, not mbedTLS.

Do you think that could be added too at the same time? I don't know how easy it is for non-Linux to load up system certs, but apparently for Linux it seems relatively easy.

Faless commented:
@akien-mga , mmhh, we'll need specific per-platform support.
mbedTLS do not handle system certificates, because it's just an SSL/TLS protocol implementation.

For this reason, we would need to read it using system calls, or read system files.
On Linux (at least Debian and derivatives) there's usually `/etc/ssl/certs/ca-certificates.crt` which we could read, but I'm not sure if that's the "standard". We could try a best effort approach I guess.

On Mac/Windows/Android/iOS there are likely dedicated functions we can call.

akien-mga commented:
Thanks, that's probably out of scope of this PR then, I'll open another issue about it.

akien-mga commented:
Do we have support for setting a custom editor certs bundle? This way distros could build with `builtin_certs=no` and then patch EditorSettings to default to their bundle's location (at least until we provide better "out of the box" support for locating their certs).

akien-mga commented on SConstruct:
```diff
@@ -190,6 +190,7 @@ opts.Add(BoolVariable('builtin_squish', "Use the built-in squish library", True)
 opts.Add(BoolVariable('builtin_thekla_atlas', "Use the built-in thekla_altas library", True))
 opts.Add(BoolVariable('builtin_zlib', "Use the built-in zlib library", True))
 opts.Add(BoolVariable('builtin_zstd', "Use the built-in Zstd library", True))
+opts.Add(BoolVariable('builtin_certs', "Bundle default SSL certificates to be used if you don't specify an override in the project settings", True))
```
Those options are listed alphabetically ;)

LikeLakers2 replied:
Butting in to say that I don't think that option fits in this section anyways -- this is for third-party libraries, and certs aren't a library in the same sense.

Faless committed:
Allow system certs file to be used by Editor.

Note, it will only used by the Editor, not when running the game.
This allows package maintainer to compile Godot to use system installed
certificates when accessing the AssetLib.

Faless commented:
@akien-mga , it's a bit hacky, but this allows package maintainers to build Godot versions where the `Editor` uses a custom defined path as source for SSL certificates.
NOTE: This only apply to the `Editor`, meaning, when you press play you will get the usual behaviour of using ProjectSettings or Builtin (even if `TOOLS_ENABLED`).

uldall commented:
@akien-mga Did you ever come around to creating an issue about having exported games use the system provided certs?

Faless commented:
@uldall #17328

Faless commented:
@uldall , no, I don't think there is one. Sadly, it seems it requires quite some work on the platform compatibility side. But feel free to open one
